File: frontend/src/types/ontology.ts
export interface OntologyStats {
  class_count: number
  individual_count: number
  property_count: number
  triple_count: number
}

export interface Ontology {
  id: string
  name: string
  description?: string
  base_iri: string
  version?: string
  created_at: string
  updated_at: string
  stats: OntologyStats
}

export interface OntologyCreate {
  name: string
  description?: string
  base_iri: string
  version?: string
}

export interface OntologyUpdate {
  name?: string
  description?: string
  version?: string
}

export interface PaginatedResponse<T> {
  items: T[]
  total: number
  page: number
  page_size: number
  has_next: boolean
}

export interface ErrorResponse {
  error: string
  detail?: string
  code?: string
}

export interface JobResponse {
  job_id: string
  status: 'pending' | 'running' | 'completed' | 'failed'
  message?: string
  result?: Record<string, unknown>
}

File: backend/models/ontology.py
"""
Ontology Pydantic models
"""
from __future__ import annotations

from datetime import datetime
from typing import Generic, Literal, TypeVar

from pydantic import BaseModel, Field

T = TypeVar("T")


class OntologyStats(BaseModel):
    concepts: int = 0
    individuals: int = 0
    object_properties: int = 0
    data_properties: int = 0
    named_graphs: int = 0


class Ontology(BaseModel):
    id: str
    iri: str
    label: str
    description: str | None = None
    version: str | None = None
    created_at: datetime
    updated_at: datetime
    stats: OntologyStats = Field(default_factory=OntologyStats)


class OntologyCreate(BaseModel):
    iri: str
    label: str
    description: str | None = None
    version: str | None = None


class OntologyUpdate(BaseModel):
    label: str | None = None
    description: str | None = None
    version: str | None = None


class PaginatedResponse(BaseModel, Generic[T]):
    items: list[T]
    total: int
    page: int
    page_size: int
    has_next: bool = False


class ErrorResponse(BaseModel):
    code: str
    message: str
    detail: object | None = None


class JobResponse(BaseModel):
    job_id: str
    status: Literal["pending", "running", "completed", "failed"]
    created_at: str   # ISO 8601
    completed_at: str | None = None
    result: object | None = None
    error: str | None = None
